File: solutions/p6-two-num-with-odd-occurrences/python-solution.py
from collections import Counter

# Function to find two numbers with odd occurrences in the array
def two_numbers_with_odd_occurrences(arr):
    try:
        # Step 1: Count the occurrences of each number using Counter
        # Counter creates a dictionary where keys are array elements, and values are their counts
        occurrence = Counter(arr)

        # Step 2: Filter out numbers that have odd counts
        # We use a list comprehension to collect numbers that occur an odd number of times
        result = [num for num, count in occurrence.items() if count % 2 != 0]

        # Step 3: Validate the result
        # Ensure there are exactly two numbers with odd occurrences
        if len(result) != 2:
            # If not, raise an error with a meaningful message
            raise ValueError("The array must have exactly two numbers with odd occurrences.")

        # Step 4: Return the result
        return result
    except Exception as e:
        # Handle any errors and print them out for the user
        print(f"Error: {e}")
        # Return an empty list if something went wrong
        return []

# Main section of the program
# Define the input array
array = [4, 7, 2, 7, 4, 4, 5, 2, 5, 9]

# Call the function and store the result
result = two_numbers_with_odd_occurrences(array)

# Display the result if we found it
if result:
    print("Numbers with odd occurrences:", result)

File: solutions/p7-majority-elements-in-array/python-solution.cpp
def find_majority_element(nums):
    try:
        # Step 1: Find the candidate for majority
        candidate, count = None, 0
        for num in nums:
            if count == 0:
                candidate = num
            count += (1 if num == candidate else -1)

        # Step 2: Verify if the candidate is a majority element
        count = 0
        for num in nums:
            if num == candidate:
                count += 1

        if count > len(nums) // 2:
            return candidate
        else:
            raise ValueError("No majority element found.")

    except Exception as e:
        return str(e)


# Main execution block
nums = [3, 3, 4, 2, 3, 3, 3]
result = find_majority_element(nums)
print("Majority Element:", result)
